docs/tpm-spi-dev.md

```diff
@@ -0,0 +1,368 @@
+
+One of the path to move forward is to port the `ms-tpm-20-ref` stack to some
+RTOS (such as [Zephyr](https://zephyrproject.org/)) to gain some more
+flexibility in tems of hardware switching in the current global situation.
+
+## Prototyping TPM SPI communication with Zephyr
+
+Further SPI work has been done on Zephyr, the code is available
+[here](https://github.com/lpn-plant/zephyr-spi-app). This code tests
+communication with a dummy TPM implementation as `ms-tpm-ref-20` has not been
+fully ported to Zephyr yet.
+
+Zephyr has it's own SPI driver, which still uses low level HAL API, however its
+behaviour is different from high level HAL API we were using previously. First,
+receiving of multiple bytes at once works fine, which simplifies code - instead
+of filling RX buffer one byte at a time we can do this in 2 transfers
+
+- 4 byte TPM header which contains: mode (read or write),
+  target register address and transfer size
+- data read or write transfer with matching size as defined in TPM header
+
+> Note: Current SPI implementation uses different header format than defined by
+> TPM spec. Mode is not present and, on read,  TPM always sends full register,
+> ignoring header size field.
+
+| Byte | Usage            |
+| ---- | ---------------- |
+| 0    | Size of transfer |
+| 1    | Register address |
+| 2    | Unused (zero)    |
+| 3    | Unused (zero)    |
+
+Currently, there are 3 registers (mininum required to send command and receive
+response):
+
+- TPM_REG_CMD - register used for sending command, there is no FIFO, entire
+  command is executed immediatelly after transfer is complete
+- TPM_REG_STATUS - status register, currently contains only response size
+- TPM_REG_RESP - response, there is no FIFO so response must be read during a
+  single transfer
+
+## Running Zephyr sample
+
+Zephyr sample communicates SPI1 (there problems in getting SPI2 to work under
+Zephyr). Following pins are used:
+
+| Pin | Function |
+| --- | -------- |
+| PA5 | SCK      |
+| PA6 | MISO     |
+| PA7 | MOSI     |
+
+> Note: Zephyr uses GPIO chip select on SPI1 which works only for master, not
+> slave.
+
+![](/images/tpm_spi1_con.jpg)
+
+To run Zephyr sample, use the following commands:
+
+```shell
+$ docker run -ti \
+    --privileged \
+    -v /dev:/dev \
+    -u $(id -u) \
+    -v $PWD:/workdir \
+    -v $(dirname $SSH_AUTH_SOCK):$(dirname $SSH_AUTH_SOCK) \
+    -e SSH_AUTH_SOCK=$SSH_AUTH_SOCK  \
+    ghcr.io/zephyrproject-rtos/zephyr-build:v0.23.3
+(docker)$ west init -m git@github.com:lpn-plant/zephyr-spi-app.git --mr spi zephyr-spi-workspace
+(docker)$ cd zephyr-spi-workspace
+(docker)$ west update
+(docker)$ export ZEPHYR_BASE=$PWD/zephyr
+(docker)$ west build -p auto -b nucleo_l476rg -s zephyr-spi-app/app/
+(docker)$ west flash
+```
+
+> Note: Zephyr outputs to emulated UART over USB which should be visible as
+> /dev/ttyACM*
+
+### Testing SPI communication
+
+We have provided `tpmctl` available in
```

Comment:
It is a `.c` file. Let's add short explaination on how to builid it.

Reply:
Done